ddtrace/internal/processor.py

```diff
@@ -12,18 +15,80 @@
 
 
 @attr.s
-class TraceProcessor(object):
-    _filters = attr.ib(type=List[TraceFilter])
+class Processor(six.with_metaclass(abc.ABCMeta)):
+    """A Processor is used to process spans as they are created and finished by a tracer."""
+
+    def __attrs_post_init__(self):
+        # type () -> None
+        """Default post initializer which logs the representation of the
+        Processor at the ``logging.DEBUG`` level.
+
+        The representation can be modified with the ``repr`` argument to the
+        attrs attribute::
+
+            @attr.s
+            class MyProcessor(Processor):
+                field_to_include = attr.ib(repr=True)
+                field_to_exclude = attr.ib(repr=False)
+        """
+        log.debug("initialized processor %r", self)
+
+    @abc.abstractmethod
+    def on_span_start(self, span):
+        # type: (Span) -> None
+        """Called when a span is started.
+
+        This method is useful for making upfront decisions on spans.
+
+        For example, a sampling decision can be made when the span is created
+        based on its resource name.
+        """
+        pass
+
+    @abc.abstractmethod
+    def on_span_finish(self, data):
+        # type: (Any) -> Any
+        """Called with the result of any previous processors or initially with
+        the finishing span when a span finishes.
+
+        It can return any data which will be passed to any processors that are
+        applied afterwards.
+        """
+        pass
+
+
+@attr.s(eq=False)
+class TraceFiltersProcessor(Processor):
+    """Processor for applying a list of ``TraceFilter`` to traces.
+
+    The filters are applied sequentially with the result of one passed on to
+    the next. If ``None`` is returned by a filter then execution short-circuits
+    and ``None`` is returned.
+
+    If a filter raises an exception it will be logged at the ``logging.ERROR``
+    level and the filter chain will continue with the next filter in the list.
+    """
+
+    _filters = attr.ib(type=List[TraceFilter], repr=True)
+
+    def on_span_start(self, span):
+        # type: (Span) -> None
+        return None
+
+    def on_span_finish(self, trace):
+        # type: (Optional[List[Span]]) -> Optional[List[Span]]
+        if trace is None:
+            return
 
-    def process(self, trace):
-        # type: (List[Span]) -> Optional[List[Span]]
         for filtr in self._filters:
```

Comment:
Is there any reason to not have the filters to be themselves `Processor` directly?

Reply:
This is going further towards the middleware/pipeline approach @Kyle-Verhoog and I discussed a few weeks back. The one issue that came up in that chat was ordering. While this doesn't matter for using this abstraction for filters, if we also did the same for sampling we could potentially have an ordering issue if a user doesn't put processors in the correct order.

Reply:
I don't think users should be in charge of ordering things themselves indeed. It's up to us to add another user-friendly configuration layer on top of that `Processor` mechanism.

Is there anything technically preventing this though?

Reply:
> Is there any reason to not have the filters to be themselves Processor directly?

We could translate `TraceFilter`s into `Processor`s here - I think it'd just be a little messy in `configure` to do the conversion and adding them to the processors list.

> Is there anything technically preventing this though?

No, I don't think so. We just need to come up with a public API that makes sense for users.

Reply:
Let's do this?

Keep in mind that users are us in 99% of the case, we're just exposing this from `Tracer` with high-level knobs. Unless somebody wants to build their own `Tracer` or provide their own list of `Processor`, which is unlikely.

Reply:
done!